source/common/stream_info/utility.cc

```diff
@@ -43,6 +49,240 @@ absl::optional<ResponseFlag> ResponseFlagUtils::toResponseFlag(absl::string_view
+    return 502;
+  case ProxyStatusError::TlsCertificateError:
+    return 502;
+  case ProxyStatusError::TlsAlertReceived:
+    return 502;
+  case ProxyStatusError::HttpRequestDenied:
+    return 403;
+  case ProxyStatusError::HttpResponseIncomplete:
+    return 502;
+  case ProxyStatusError::HttpResponseHeaderSectionSize:
+    return 502;
+  case ProxyStatusError::HttpResponseHeaderSize:
+    return 502;
+  case ProxyStatusError::HttpResponseBodySize:
+    return 502;
+  case ProxyStatusError::HttpResponseTrailerSectionSize:
+    return 502;
+  case ProxyStatusError::HttpResponseTrailerSize:
+    return 502;
+  case ProxyStatusError::HttpResponseTransferCoding:
+    return 502;
+  case ProxyStatusError::HttpResponseContentCoding:
+    return 502;
+  case ProxyStatusError::HttpResponseTimeout:
+    return 504;
+  case ProxyStatusError::HttpUpgradeFailed:
+    return 502;
+  case ProxyStatusError::HttpProtocolError:
+    return 502;
+  case ProxyStatusError::ProxyInternalError:
+    return 500;
+  case ProxyStatusError::ProxyConfigurationError:
+    return 500;
+  case ProxyStatusError::ProxyLoopDetected:
+    return 502;
+  case ProxyStatusError::ProxyInternalResponse:
+  case ProxyStatusError::HttpRequestError:
+  default:
+    return absl::nullopt;
+  }
+}
+
+const std::string
+ProxyStatusUtils::toString(const StreamInfo& stream_info, const ProxyStatusError error,
+                           absl::string_view server_name,
+                           const envoy::extensions::filters::network::http_connection_manager::v3::
+                               HttpConnectionManager::ProxyStatusConfig& proxy_status_config) {
+  std::vector<std::string> retval = {};
+
+  switch (proxy_status_config.proxy_name()) {
+  case envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager::
+      ProxyStatusConfig::SERVER_NAME: {
+    retval.push_back(std::string(server_name));
+    break;
+  }
+  case envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager::
+      ProxyStatusConfig::ENVOY_LITERAL:
+  default: {
+    retval.push_back(Envoy::Http::DefaultServerString::get());
+    break;
+  }
+  }
+
+  retval.push_back(absl::StrFormat("error=%s", proxyStatusErrorToString(error)));
+
+  if (!proxy_status_config.remove_details() && stream_info.responseCodeDetails().has_value()) {
+    std::vector<std::string> details = {};
+    details.push_back(stream_info.responseCodeDetails().value());
+    if (!proxy_status_config.remove_connection_termination_details() &&
+        stream_info.connectionTerminationDetails().has_value()) {
+      details.push_back(stream_info.connectionTerminationDetails().value());
+    }
+    if (!proxy_status_config.remove_response_flags() && stream_info.hasAnyResponseFlag()) {
+      details.push_back(ResponseFlagUtils::toShortString(stream_info));
+    }
+    retval.push_back(absl::StrFormat("details='%s'", absl::StrJoin(details, "; ")));
```

Comment:
Just noticed. details is sf-string and needs to be in double quotes https://datatracker.ietf.org/doc/html/rfc8941#section-3.3.3

I do not think any of our details strings has double quotes in them. So perhaps we do not need to check and escape.